src/ips/algorithm.rs

```diff
@@ -77,30 +80,64 @@ impl Ips {
 
     /// Generate peer list - main function with The Algorithm
     pub async fn generate(&mut self, state: &CrunchyState) -> Vec<Peer> {
+        // Set up logging
+        let output = match self.config.log_path {
+            Some(ref path) => File::create(path).map(|f| Box::new(f) as Box<dyn Write>),
+            None => Ok(Box::new(io::stdout()) as Box<dyn Write>),
+        };
+
+        let mut output = match output {
+            Ok(output) => output,
+            Err(e) => {
+                println!("Failed to open log file: {}", e);
+                Box::new(io::stdout()) as Box<dyn Write>
+            }
+        };
+
         // Sanity check that each node is really connected to its peers and the peers also
         // have the node in their connections.
+        writeln!(output, "IPS algorithm started... ").unwrap();
+        let start_time = std::time::Instant::now();
+
+        writeln!(output, "Checking for nodes connected to themselves...").unwrap();
         for (idx, node) in state.nodes.iter().enumerate() {
             if node.connections.contains(&idx) {
-                println!("{} is connected to itself.", node.addr);
+                writeln!(output, "{} is connected to itself.", node.addr).unwrap();
```

Comment:
`f` just because we repeat this call too many times, here's a suggestion to rename output to `f` as in `file`
or `o` as in `output`

one can easily check what it means at the function start - but usually one assumes it's either a file or 
 the stdout
```suggestion
                writeln!(f, "{} is connected to itself.", node.addr).unwrap();
```

Reply:
Changed to `o` - feel better when it's `o` than `f` as this could suggest it's always some file.